### ISSUE TITLE
[HOLD for payment 2024-08-23] [$2000] Add support for copy/pasting images on iOS

### ISSUE THREAD

**Issue by roryabraham** [Expensify/App#41239]
slack proposal: https://expensify.slack.com/archives/C01GTK53T8Q/p1714239125412939

### Problem
One of the most common flows for sharing images on iOS is to copy the image somewhere and paste it somewhere else, and NewDot does not support that flow. For example, I wanted to share an image I found on Google in New Expensify. The first thing I did was copy the image to my clipboard, but when I long-pressed on the composer to try and paste that image, the only option available was AutoFill.
This left me with two more less desirable options:

- copy the URL to the image, then write that URL in a markdown comment with inline image syntax (this was the next thing I tried, and it actually didn't work for the URL I had)
- save the image to my photos, then upload it to New Expensify. This is less ideal because it pollutes my photos gallery with something I don't want to keep there.

This was a very basic flow that did not work very well at all, and it was a frustrating experience. It's table-stakes for any chat app.

### Solution
Implement support for pasting images on iOS.
<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~0195bbb40518505764</li>
        <li>Upwork Job ID: 1790820991593271296</li>
        <li>Last Price Increase: 2024-07-16</li>
<li>Automatic offers: </li>
<ul>
<li>ahmedGaber93 | Reviewer | 103152998</li>
</ul></li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @sonialiap</details>

**Comment by roryabraham** [comment]
I don't think we need any design review here

**Comment by dubielzyk-expensify** [comment]
Love this ticket. As for the solution, I kinda find it weird to trigger the "paste" permission on `(+)` click. Does this only work when you have copied a picture? What happens if you copy text? 

In other apps, and even for Expensify Web and MacOS, you can paste a picture into the actual compose bar text input which to me feels more natural than to click the `(+)` button. Thoughts?  cc @Expensify/design

**Comment by trjExpensify** [comment]
🤔 Also, what about Android to maintain cross-platform consistency given @Beamanator's recent discovery [here](https://github.com/Expensify/App/issues/4629#issuecomment-2058589631) on Rory's OG issue for this:

>I did figure out that on an android device, you can go to chrome, search for an image, press & hold, and copy to the clipboard -> so shouldn't we make sure this is possible on our android app too? 👍

**Comment by shawnborton** [comment]
100% agree, I find it super bizarre to override the default `+` behavior here. We should definitely not do that. 

Why can't you just paste into the compose box as if you were pasting text, and then that triggers the paste dialog?

**Comment by shawnborton** [comment]
Well if a link was on your clipboard, I would think you would just paste the link. But if an image was on your clipboard, then it makes sense to trigger the modal window.

**Comment by ahmedGaber93** [comment]
@ShridharGoel Thanks for the proposal.

I think pasting the image from the URL in the clipboard seem a workaround, because when we copied the image, we click "copy image" not "copy image address", so we may need to handle the image as fileObject or base64 string.

**Comment by dubielzyk-expensify** [comment]
The paste to compose bar is how most apps on iOS works. It works like this on Messages, Telegram, and Signal. So I suggest we do that. Like Shawn said, let's trigger the modal window like we do on web/mac os.

**Comment by dannymcclain** [comment]
>The paste to compose bar is how most apps on iOS works. It works like this on Messages, Telegram, and Signal. So I suggest we do that. Like Shawn said, let's trigger the modal window like we do on web/mac os.

Totally agree.

**Comment by roryabraham** [comment]
Yes, paste should appear on the _native_ context menu that appears when you double-tap on the composer input.

**Comment by sonialiap** [comment]
@roryabraham should this issue be external or internal? Are we looking for contributors to build this or is thing something you're working on?

**Comment by roryabraham** [comment]
It should be external

**Comment by roryabraham** [comment]
bumping to $500 to try and get some fresh proposals

**Comment by ahmedGaber93** [comment]
Waiting on proposals

**Comment by ahmedGaber93** [comment]
Waiting on proposals

**Comment by roryabraham** [comment]
With our high focus on reliability over new features, this is kind of taking a back-seat. However, I'd say this is reasonably high priority still because it has a high impact on UX, particularly on iOS for P2P / chat uses.

So I'm going to bump the bounty.

**Comment by eh2077** [comment]
Hi @roryabraham , we had solid discussions and POC from an old issue which was closed without moving forward at that time.

- Investigations see [comment1](https://github.com/Expensify/App/issues/36909#issuecomment-1978754442) and [comment2](https://github.com/Expensify/App/issues/36909#issuecomment-1980229916)
- POC demo see [link](https://github.com/Expensify/App/issues/36909#issuecomment-1984173728)

I think we can continue with previous discussions and bring the feature described in this issue to the App. That said, I'm happy to continue work on it with @kacper-mikolajczak if possible.

Cc @kacper-mikolajczak @mountiny

**Comment by samilabud** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Add support for copy/pasting images on iOS

### What is the root cause of that problem?
Create new feature in IOS

### What changes do you think we should make in order to solve the problem?
We can add a new option in the [attachment picker](https://github.com/Expensify/App/blob/main/src/pages/home/report/ReportActionCompose/AttachmentPickerWithMenuItems.tsx), maybe inside the moneyRequestOptions like:

```
if (Platform.OS === 'ios') {
            const newOptions = {
                [CONST.IOU.TYPE.SUBMIT]: {
                    icon: Expensicons.Copy,
                    text: 'Paste image',
                    // eslint-disable-next-line @typescript-eslint/no-misused-promises
                    onSelected: () => checkClipboardForImage(),
                },
            };
            options = {
                ...options,
                ...newOptions,
            };
        }
```
Then we should install two new components/libraries ([react-native-fs](https://www.npmjs.com/package/react-native-fs) and [react-native-clipboard-plus](https://www.npmjs.com/package/react-native-clipboard-plus)), these are needed to make the copy and pasted in Native Platforms. 

To handle the paste method we can create the next method:
```
const checkClipboardForImage = async () => {
        if (Platform.OS !== 'ios') {
            return;
        }
        const clipboardContent = await ClipboardPlus.paste();
        if (clipboardContent) {
            if (clipboardContent.image) {
                const imageBase64 = `data:image/png;base64,${clipboardContent.image.toString()}`;
                const imageFile = await FileUtilsIOS.base64ToFileNative(imageBase64, 'image.png');
                displayFileInModal(imageFile, imageBase64);
            }
        }
    };
```
This last one use a new library created by us called FileUtilsIOS (see [draft PR](https://github.com/Expensify/App/pull/42595) for details).

Finally we shoud modify the [validateAndDisplayFileToUpload](https://github.com/Expensify/App/blob/8375abea3a377e7416404addde255b6ac8530a9b/src/components/AttachmentModal.tsx#L307) method to accept the base64Image like:
  ```
const validateAndDisplayFileToUpload = useCallback(
        (data: FileObject, imageBase64 = '') => {
            if (!data || !isDirectoryCheck(data)) {
                return;
            }
            let fileObject = data;
            if ('getAsFile' in data && typeof data.getAsFile === 'function') {
                fileObject = data.getAsFile();
            }
            if (!fileObject) {
                return;
            }

            if (!isValidFile(fileObject)) {
                return;
            }
            if (imageBase64) {
                setIsModalOpen(true);
                setSourceState(imageBase64);
                setFile(fileObject);
                setModalType('centered_unswipeable');
            } else if (fileObject instanceof File) {
.
.
.//same code
```

So this will be the result:

https://github.com/Expensify/App/assets/5216036/b1442da2-be1d-496f-9f82-26a6f2b98f32


[Draft PR](https://github.com/Expensify/App/pull/42595)

**Comment by ahmedGaber93** [comment]
@samilabud Thanks for the proposal.
> We can add a new option in the [attachment picker](https://github.com/Expensify/App/blob/main/src/pages/home/report/ReportActionCompose/AttachmentPickerWithMenuItems.tsx), maybe inside the moneyRequestOptions like:


I think we don't need this behavior, we need to handle the normal case when long/double click on the input then click paste. Also, [react-native-clipboard-plus](https://www.npmjs.com/package/react-native-clipboard-plus) seem low popularity and maintenance

**Comment by ahmedGaber93** [comment]
@suneox Thanks for the proposal.

I see your proposal try to handle base64 and JPG images, did you test it in other formats? And can you share a test branch and the formats you tested it?

**Comment by samilabud** [comment]
Hi @ahmedGaber93,

> we need to handle the normal case when long/double click on the input then click paste

 As far I can see in React Native the TextInput component does not directly support pasting images, I mean there are not event that we can use to handle it at least the pasted content be a text/string.

Please let me know if this makes sense to you, if so I'll change my proposal using another library to handle the clipboard in IOS that has more popularity.

**Comment by ahmedGaber93** [comment]
> Please let me know if this makes sense to you

@samilabud it recommended by design team here https://github.com/Expensify/App/issues/41239#issuecomment-2084636648 to paste it in composer directly

**Comment by ahmedGaber93** [comment]
@suneox related to this problem and your fix for it
### The problem
When user copy image from iPhone photos app, and try to paste it in composer, the paste option will not be available, and the only option available was AutoFill.

### Your fix
- listen to composer `onPressIn` and check if [Clipboard.hasImage()](https://github.com/react-native-clipboard/clipboard/blob/21c5f423b76bee9e75087423262132d1c43c4612/src/Clipboard.ts#L130), and if it has, catch it as PNG in base64 by [Clipboard.getImagePNG()](https://github.com/react-native-clipboard/clipboard/blob/21c5f423b76bee9e75087423262132d1c43c4612/src/Clipboard.ts#L42) and save it in react.ref. Use `Clipboard.setString(EXP_IMAGE)` to copy random text `EXP_IMAGE` to clipboard which make the paste option appear.

- listen to composer `onChangeText`and when user paste, if the text includes the copied text `EXP_IMAGE` that's mean he pasted image. Use image ref and handle paste file.

___
 
 ### Test results
 
- calling `getImagePNG` on `onPressIn` will trigger iOS to request paste permission, and this seems weird because the user just press on the composer not click paste item yet.

- If we try to paste JPG image from iPhone photos app it will work, but from Google images it not works and instead pasted link

- Also, I am wonder why `hasImage` and `getImagePNG` are not documented on Clipboard lib docs.

**Comment by ahmedGaber93** [comment]
@suneox for the alternative solution, what is the test result of it? 
for example, and not as a limitation

- [ ] confirm paste option is appeared when you copy image from local (Photo app)
- [ ] confirm you are able to paste image from local (Photo app)
- [ ] confirm you are able to paste image from Google images/web
- [ ] confirm you are able to paste base64 images
- [ ] what formats are you able to paste 
    - [ ] png
    - [ ] jpg/jpeg
    - [ ] gif
    - [ ] tiff

**Comment by ahmedGaber93** [comment]
> In this case the IOS copy the URL is expected result

I also thought that, but when I tested it on Messages app, it works fine.

https://github.com/Expensify/App/assets/41129870/04b77ba3-3c49-42d4-8200-1fd00d0dd581

**Comment by ahmedGaber93** [comment]
Not overdue, It's the weekend.

**Comment by ahmedGaber93** [comment]
@suneox I still can't paste image from web, see video here https://github.com/Expensify/App/issues/41239#issuecomment-2138673462

**Comment by ahmedGaber93** [comment]
> Ah i see we have to handle this case with image type is com.apple.uikit.image

@suneox Can you handle it in the test branch.

**Comment by ahmedGaber93** [comment]
Still review @suneox proposal

**Comment by ahmedGaber93** [comment]
@suneox Can you share the images links you test with it? All images from web not works with me.

**Comment by ahmedGaber93** [comment]
All web images not works with me in the test branch, include both links here https://github.com/Expensify/App/issues/41239#issuecomment-2166157153

**Comment by ahmedGaber93** [comment]
### Proposals summary right now
 ___
 
1️⃣ - [proposal](https://github.com/Expensify/App/issues/36909#issuecomment-1984173728) in the second issue that use [react-native-paste-input](https://github.com/mattermost/react-native-paste-input) 

#### test result
- [x] paste option is appearing correctly and can paste image from photos app
- [ ] paste image from web as image, not as a link

this proposal require an external lib, this lib wrap react-native `TextInput` and add some logic around it beside the Native logic, and we need to import input from this lib instead of reaact-native
```javascript
import PasteInput from "@mattermost/react-native-paste-input";
```

- This lib also supports android.
___

2️⃣ - @suneox's [alternative solution](https://github.com/Expensify/App/issues/41239#issuecomment-2130158574). The test branch [here](https://github.com/suneox/RNTest-p/tree/ios-handle-paste-image)

#### test result
- [x] paste option is appearing correctly and can paste image from photos app
- [ ] paste image from web as image, not as a link

this solution suggest edit `MarkdownTextInput` and listen on paste option in `RCTBaseTextInputView+Markdown.mm`, this solution idea is close to react-native-paste-input but add the changes locally in `MarkdownTextInput` instead of external lib

- If we will support android, we will need to update `MarkdownTextInput` in android files also

___

The both proposals not cover paste image from web like messages app and paste images as a link, I think we need to support web images before move forward with any of them.

**Comment by ahmedGaber93** [comment]
@suneox Related to your main proposals, I think your way to detect `onPasteFile` and the way to make paste option appear after copy image from gallery is not professional, and I think it not work as expected.

If `Clipboard.getImagePNG()` work for iOS and `Clipboard.getImage()` work for android without permissions, I think if we can detect `onPasteFile` by native in `MarkdownTextInput` for both android and iOS, and in E/App use Clipboard to get the image content. What do you think?

**Comment by ahmedGaber93** [comment]
Still open for proposal that cover both gallery and web images

**Comment by ahmedGaber93** [comment]
Still open for proposal that cover both gallery and web images

**Comment by ahmedGaber93** [comment]
Still open for proposal that cover both gallery and web images

**Comment by ahmedGaber93** [comment]
Still open for proposals

**Comment by ahmedGaber93** [comment]
Waiting on proposals

**Comment by ahmedGaber93** [comment]
Waiting on proposals

**Comment by ahmedGaber93** [comment]
Waiting on proposals

**Comment by roryabraham** [comment]
Let's double this

**Comment by s77rt** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Add image copy/paste support for Composer on iOS

### What is the root cause of that problem?
None. This is a feature request and not a bug

### What changes do you think we should make in order to solve the problem?
Extend `RCTUITextView` in RN via a patch and do the following:
1. Override [`canPerformAction:`](https://github.com/facebook/react-native/blob/36a4a2450144228e8a3ccc63541428751a7e4555/packages/react-native/Libraries/Text/TextInput/Multiline/RCTUITextView.mm#L255) method to force show the `Paste` option if the clipboard has an image
2. Override [`paste:`](https://github.com/facebook/react-native/blob/36a4a2450144228e8a3ccc63541428751a7e4555/packages/react-native/Libraries/Text/TextInput/Multiline/RCTUITextView.mm#L166) method to call `textInputDidPaste` method
3. Implement `textInputDidPaste`; this method will just send an event to the component view via `onPaste` prop
4. In App (`Composer`)  add `onPaste` prop and call [`Clipboard.getImagePNG`](https://github.com/react-native-clipboard/clipboard/blob/21c5f423b76bee9e75087423262132d1c43c4612/ios/RNCClipboard.mm#L105)

### What alternative solutions did you explore? (Optional)
None but there are variations of the proposed solution e.g. instead of using onPaste + Clipboard.getImagePNG an alternative is to implement onPasteImage in which the image is provided in the event callback

**Comment by s77rt** [comment]
https://github.com/Expensify/App/assets/16493223/a13454af-a192-4260-b66e-55dd4f4176d1

**Comment by ahmedGaber93** [comment]
@s77rt Can you share a test branch? Also are we able to use the same approach in android, Clipboard has `Clipboard.getImage()` that to work for android

**Comment by s77rt** [comment]
@ahmedGaber93

> Can you share a test branch?

I don't have a test branch yet. I will see if I can provide one. Hopefully that's not a blocker.

> Also are we able to use the same approach in android

Yes, we can use the same approach. I have updated the [proposal](https://github.com/Expensify/App/issues/41239#issuecomment-2223319480)

**Comment by roryabraham** [comment]
@s77rt I'm excited to see your proposal works, but I don't want a solution that depends long-term on any patches. So if possible, let's write up a more detailed proposal to share with Meta for feedback. This proposal should include considerations for testing in the upstream repo (minimally using rn-tester, but ideally including some automated tests).

**Comment by s77rt** [comment]
@roryabraham I have raised an upstream PR https://github.com/facebook/react-native/pull/45425 that adds the `onPaste` prop to `TextInput` covering most of the above changes. Except that in iOS the Paste option won't show still if all we have in the clipboard is an image. The reason behind this is that image pasting is app dependent i.e. in our case we will show the attachment modal but in others (e.g. in RNTester) nothing will happen as there is no handler for image pasting. This leaves us with a patch but a minor one.

**Comment by ahmedGaber93** [comment]
Not overdue, I think we should remove Daily and add Weekly label here.

**Comment by roryabraham** [comment]
@s77rt awesome to see that PR already imported!

**Comment by s77rt** [comment]
I assume we'd go with a patch for now until we upgrade to the new RN version? cc @roryabraham

**Comment by roryabraham** [comment]
> I assume we'd go with a patch for now until we upgrade to the new RN version?

Yes, we can do that. That hopefully helps give Meta some confidence that we're testing this in production (though we don't have as many users as some of their other partners by any means)

**Comment by s77rt** [comment]
Added the `onPaste` feature proposal https://github.com/react-native-community/discussions-and-proposals/pull/804

**Comment by s77rt** [comment]
Raised PR https://github.com/Expensify/App/pull/45722

**Comment by ahmedGaber93** [comment]
Not overdue PR in progress

**Comment by s77rt** [comment]
**Update:** Waiting on https://github.com/react-native-clipboard/clipboard/pull/253. This PR adds `getImage` support for iOS fixing [a performance issue](https://github.com/Expensify/App/pull/45722#issuecomment-2242445052).

**Comment by roryabraham** [comment]
The guidance here should be to try and merge a patch that's as similar to the upstream code as possible. Looking at https://github.com/react-native-community/discussions-and-proposals/pull/804 it sounds like they'd like you to try to match web APIs more closely with your implementation, including binding some clipboard data in the `onPaste` event. I see you asked if we can address this in a follow-up, but the idea with adding a patch is that we are first and foremost making a good-faith effort to improve the project upstream, and the patch just expedites things so we don't have to wait as long for review, release, and upgrade. In this case, they've provided feedback, so I think it would be better to try and address that feedback before moving forward with a different implementation

**Comment by ahmedGaber93** [comment]
Not overdue, we are waiting for https://github.com/Expensify/App/issues/41239#issuecomment-2242775101

**Comment by s77rt** [comment]
We are no longer waiting on `react-native-clipboard`. The `onPaste` callback event now contains the clipboard data which we can use directly. The https://github.com/Expensify/App/pull/45722 PR is basically ready for merge but waiting on RN team feedback on the proposal https://github.com/react-native-community/discussions-and-proposals/pull/804

**Comment by ahmedGaber93** [comment]
Not overdue, still waiting https://github.com/Expensify/App/issues/41239#issuecomment-2254486512

**Comment by ahmedGaber93** [comment]
@ishpaul777, thanks for the help. Nice to share 25% of C+ bounty with him for reviewing edge case in real iOS device.

**Comment by sonialiap** [comment]
Payment summary:
- @ahmedGaber93 $1500 primary C+ - offer in [upwork](https://www.upwork.com/ab/applicants/1790820991593271296/hired)
- @ishpaul777 $500 review help - offer in [upwork](https://www.upwork.com/ab/applicants/1824424909495395881/hired)
- @s77rt $2000 - please request in ND

**Comment by sonialiap** [comment]
@stephanieelliott I'm OOO Aug 19-30, adding leave buddy. 
Status: if no regressions then please complete payments summarized [here](https://github.com/Expensify/App/issues/41239#issuecomment-2293433932)

**Comment by s77rt** [comment]
Payment not due yet. Waiting for https://github.com/Expensify/App/pull/47203 to hit production (+ 7 days)

**Comment by stephanieelliott** [comment]
Not overdue, holding this for payment on the 23rd.

**Comment by roryabraham** [comment]
Ready to pay

**Comment by s77rt** [comment]
Can we not apply the regressions penalties here due to complexity?

**Comment by roryabraham** [comment]
Yes, that seems reasonable in this case, as long as you fix regressions that do come up. I see you created another PR already, so thank you for that.

**Comment by s77rt** [comment]
False positive ^

**Comment by stephanieelliott** [comment]
OK, I will go ahead and pay this now seeing as we're not going to adjust for regressions. 

Summarizing payment on this issue: 

 - Contributor: @ahmedGaber93 $1000 via Upwork - PAID 
 - Contributor+: @s77rt $1000 via NewDot - please request! 
 
 Upwork job is here: https://www.upwork.com/jobs/~0195bbb40518505764

**Comment by ahmedGaber93** [comment]
@stephanieelliott payment summary was changed and summarized by @sonialiap here https://github.com/Expensify/App/issues/41239#issuecomment-2293433932 to split 25% of C+ bounty to @ishpaul777 for debugging reviewing a case on a iOS real fevice

**Comment by JmillsExpensify** [comment]
$2,000 approved for @s77rt based on [this first summary](https://github.com/Expensify/App/issues/41239#issuecomment-2293433932).

**Comment by mallenexpensify** [comment]
Paid @ishpaul777 , updated payment summary above. 

@ahmedGaber93 , requested $500 refund, per the split bounty, please confirm via a comment here once you've issued it. Thx

**Comment by ahmedGaber93** [comment]
> please confirm via a comment here once you've issued it. 

@mallenexpensify Done.

**Comment by mallenexpensify** [comment]
Thanks @ahmedGaber93 , confirming refund has been addressed. 
@s77rt do we want a regression test from here? If not, why?  Thx

**Comment by s77rt** [comment]
@mallenexpensify This is a new feature and it does require regression test.

**Regression Test Proposal**

1. Copy image (from Gallery or Google search results)
2. Open any report
3. Paste the image
4. Verify that an attachment modal is displayed
5. Upload the photo
6. Verify the photo is uploaded with success

**Comment by sonialiap** [comment]
Thanks Steph and Matt for handling payment while I was OOO!

Submitted Regression test to the QA team. I believe this is all done and ready to close
